code/game/machinery/hologram.dm

```diff
@@ -68,197 +102,307 @@ var/list/holopads = list()
 	default_deconstruction_crowbar(P)
 
 
-/obj/machinery/hologram/holopad/attack_hand(var/mob/living/carbon/human/user) //Carn: Hologram requests.
+/obj/machinery/hologram/holopad/attack_hand(mob/living/carbon/human/user)
+	if(..())
+		return
+
+	if(outgoing_call)
+		return
+
+	user.set_machine(src)
+	interact(user)
+
+/obj/machinery/hologram/holopad/AltClick(mob/living/carbon/human/user)
+	if(..())
+		return
+	if(isAI(user))
+		hangup_all_calls()
+		return
+
+//Stop ringing the AI!!
+/obj/machinery/hologram/holopad/proc/hangup_all_calls()
+	for(var/I in holo_calls)
+		var/datum/holocall/HC = I
+		HC.Disconnect(src)
+
+/obj/machinery/hologram/holopad/interact(mob/living/carbon/human/user) //Carn: hologram requests.
 	if(!istype(user))
 		return
-	if(alert(user,"Would you like to request an AI's presence?",,"Yes","No") == "Yes")
-		if(last_request + 200 < world.time) //don't spam the AI with requests you jerk!
+
+	var/dat
+	if(temp)
+		dat = temp
+	else
+		dat = "<a href='?src=[UID()];AIrequest=1'>Request an AI's presence.</a><br>"
+		dat += "<a href='?src=[UID()];Holocall=1'>Call another holopad.</a><br>"
+
+		if(LAZYLEN(holo_calls))
+			dat += "=====================================================<br>"
+
+		var/one_answered_call = FALSE
+		var/one_unanswered_call = FALSE
+		for(var/I in holo_calls)
+			var/datum/holocall/HC = I
+			if(HC.connected_holopad != src)
+				dat += "<a href='?src=[UID()];connectcall=\ref[HC]'>Answer call from [get_area(HC.calling_holopad)].</a><br>"

```

Comment:
Instead of `\ref[HC]`, this should be `[HC.UID()]` - otherwise, the `locateUID` call will return `null` and cause the call to fail.